File: server/Sources/App/Support/GitHubStatsCalculator.swift
import Foundation

enum GitHubStatsCalculator {
    struct Day {
        let date: String
        let count: Int
    }

    static func peakDay(from days: [Day]) -> (count: Int, date: String?) {
        guard let best = days.max(by: { $0.count < $1.count }), best.count > 0 else {
            return (0, nil)
        }
        return (best.count, best.date)
    }

    static func longestStreak(from days: [Day]) -> Int {
        let sorted = days.sorted { $0.date < $1.date }
        var longest = 0
        var current = 0
        var previousDate: Date?

        let formatter = DateFormatter()
        formatter.calendar = Calendar(identifier: .gregorian)
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = TimeZone(secondsFromGMT: 0)
        formatter.dateFormat = "yyyy-MM-dd"

        for day in sorted where day.count > 0 {
            guard let date = formatter.date(from: day.date) else { continue }

            if let previousDate {
                let delta = Calendar.current.dateComponents([.day], from: previousDate, to: date).day ?? 0
                if delta == 1 {
                    current += 1
                } else {
                    current = 1
                }
            } else {
                current = 1
            }

            longest = max(longest, current)
            previousDate = date
        }

        return longest
    }

    static func lifetimeCommits(from days: [Day]) -> Int {
        days.reduce(0) { $0 + $1.count }
    }
}

File: server/src/github/github-stats-calculator.ts
export type ActivityDay = {
  date: string;
  count: number;
};

export function peakDay(days: ActivityDay[]): { count: number; date: string | null } {
  const best = days.reduce<ActivityDay | null>((acc, day) => {
    if (!acc || day.count > acc.count) return day;
    return acc;
  }, null);
  if (!best || best.count <= 0) return { count: 0, date: null };
  return { count: best.count, date: best.date };
}

export function longestStreak(days: ActivityDay[]): number {
  const sorted = [...days].sort((a, b) => a.date.localeCompare(b.date));
  let longest = 0;
  let current = 0;
  let previousDate: Date | null = null;

  for (const day of sorted) {
    if (day.count <= 0) continue;
    const date = new Date(`${day.date}T00:00:00Z`);
    if (Number.isNaN(date.getTime())) continue;

    if (previousDate) {
      const delta = Math.round((date.getTime() - previousDate.getTime()) / (1000 * 60 * 60 * 24));
      current = delta === 1 ? current + 1 : 1;
    } else {
      current = 1;
    }

    longest = Math.max(longest, current);
    previousDate = date;
  }

  return longest;
}